#### Next up : Implement Open AI compatible Ollama API

# YOUR FIRST LAB
### Please read this section. This is valuable to get you prepared, even if it's a long read -- it's important stuff.

### Also, be sure to read [README.md](../README.md)! More info about the updated videos in the README and [top of the course resources in purple](https://edwarddonner.com/2024/11/13/llm-engineering-resources/)

## Your first Frontier LLM Project

By the end of this course, you will have built an autonomous Agentic AI solution with 7 agents that collaborate to solve a business problem. All in good time! We will start with something smaller...

Our goal is to code a new kind of Web Browser. Give it a URL, and it will respond with a summary. The Reader's Digest of the internet!!

Before starting, you should have completed the setup linked in the README.

### If you're new to working in "Notebooks" (also known as Labs or Jupyter Lab)

Welcome to the wonderful world of Data Science experimentation! Simply click in each "cell" with code in it, such as the cell immediately below this text, and hit Shift+Return to execute that cell. Be sure to run every cell, starting at the top, in order.

Please look in the [Guides folder](../guides/01_intro.ipynb) for all the guides.

## I am here to help

If you have any problems at all, please do reach out.  
I'm available through the platform, or at ed@edwarddonner.com, or at https://www.linkedin.com/in/eddonner/ if you'd like to connect (and I love connecting!)  
And this is new to me, but I'm also trying out X at [@edwarddonner](https://x.com/edwarddonner) - if you're on X, please show me how it's done 😂  

## More troubleshooting

Please see the [troubleshooting](../setup/troubleshooting.ipynb) notebook in the setup folder to diagnose and fix common problems. At the very end of it is a diagnostics script with some useful debug info.

## If this is old hat!

If you're already comfortable with today's material, please hang in there; you can move swiftly through the first few labs - we will get much more in depth as the weeks progress. Ultimately we will fine-tune our own LLM to compete with OpenAI!

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/important.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#900;">Please read - important note</h2>
            <span style="color:#900;">The way I collaborate with you may be different to other courses you've taken. I prefer not to type code while you watch. Rather, I execute Jupyter Labs, like this, and give you an intuition for what's going on. My suggestion is that you carefully execute this yourself, <b>after</b> watching the lecture. Add print statements to understand what's going on, and then come up with your own variations. If you have a Github account, use this to showcase your variations. Not only is this essential practice, but it demonstrates your skills to others, including perhaps future clients or employers...</span>
        </td>
    </tr>
</table>
<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/resources.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#f71;">This code is a live resource - keep an eye out for my emails</h2>
            <span style="color:#f71;">I push updates to the code regularly. As people ask questions, I add more examples or improved commentary. As a result, you'll notice that the code below isn't identical to the videos. Everything from the videos is here; but I've also added better explanations and new models like DeepSeek. Consider this like an interactive book.<br/><br/>
                I try to send emails regularly with important updates related to the course. You can find this in the 'Announcements' section of Udemy in the left sidebar. You can also choose to receive my emails via your Notification Settings in Udemy. I'm respectful of your inbox and always try to add value with my emails!
            </span>
        </td>
    </tr>
</table>
<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/business.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#181;">Business value of these exercises</h2>
            <span style="color:#181;">A final thought. While I've designed these notebooks to be educational, I've also tried to make them enjoyable. We'll do fun things like have LLMs tell jokes and argue with each other. But fundamentally, my goal is to teach skills you can apply in business. I'll explain business implications as we go, and it's worth keeping this in mind: as you build experience with models and techniques, think of ways you could put this into action at work today. Please do contact me if you'd like to discuss more or if you have ideas to bounce off me.</span>
        </td>
    </tr>
</table>

### If necessary, install Cursor Extensions

1. From the View menu, select Extensions
2. Search for Python
3. Click on "Python" made by "ms-python" and select Install if not already installed
4. Search for Jupyter
5. Click on "Jupyter" made by "ms-toolsai" and select Install of not already installed


### Next Select the Kernel

Click on "Select Kernel" on the Top Right

Choose "Python Environments..."

Then choose the one that looks like `.venv (Python 3.12.x) .venv/bin/python` - it should be marked as "Recommended" and have a big star next to it.

Any problems with this? Head over to the troubleshooting.

### Note: you'll need to set the Kernel with every notebook..

In [1]:
# imports

import os
from dotenv import load_dotenv
from scraper import fetch_website_contents
from IPython.display import Markdown, display
from openai import OpenAI

from langchain_community.embeddings import OllamaEmbeddings
from langchain_community.llms import Ollama
# If you get an error running this cell, then please head over to the troubleshooting notebook!

# Connecting to OpenAI (or Ollama)

The next cell is where we load in the environment variables in your `.env` file and connect to OpenAI.  

If you'd like to use free Ollama instead, please see the README section "Free Alternative to Paid APIs", and if you're not sure how to do this, there's a full solution in the solutions folder (day1_with_ollama.ipynb).

## Troubleshooting if you have problems:

If you get a "Name Error" - have you run all cells from the top down? Head over to the Python Foundations guide for a bulletproof way to find and fix all Name Errors.

If that doesn't fix it, head over to the [troubleshooting](../setup/troubleshooting.ipynb) notebook for step by step code to identify the root cause and fix it!

Or, contact me! Message me or email ed@edwarddonner.com and we will get this to work.

Any concerns about API costs? See my notes in the README - costs should be minimal, and you can control it at every point. You can also use Ollama as a free alternative, which we discuss during Day 2.

In [2]:
# Load environment variables in a file called .env

load_dotenv(override=True)
api_key = os.getenv('OPENAI_API_KEY')

# Check the key

if not api_key:
    print("No API key was found - please head over to the troubleshooting notebook in this folder to identify & fix!")
elif not api_key.startswith("sk-proj-"):
    print("An API key was found, but it doesn't start sk-proj-; please check you're using the right key - see troubleshooting notebook")
elif api_key.strip() != api_key:
    print("An API key was found, but it looks like it might have space or tab characters at the start or end - please remove them - see troubleshooting notebook")
else:
    print("API key found and looks good so far!")


API key found and looks good so far!


# Let's make a quick call to a Frontier model to get started, as a preview!

In [3]:
# To give you a preview -- calling OpenAI with these messages is this easy. Any problems, head over to the Troubleshooting notebook.

message = "Hello, GPT! This is my first ever message to you! Hi!"

messages = [{"role": "user", "content": message}]

messages

[{'role': 'user',
  'content': 'Hello, GPT! This is my first ever message to you! Hi!'}]

In [4]:
messages = [
    {"role": "system", "content": "You are a helpful assistant"},
    {"role": "user", "content": "What is two plus two"}
]

llm = Ollama(model="llama3.1")
res = llm.invoke(input = messages)
res.split("\n")


/var/folders/p1/g_l_6ln97jx73n6xnw93ss2h0000gn/T/ipykernel_63110/3675485545.py:6: LangChainDeprecationWarning: The class `Ollama` was deprecated in LangChain 0.3.1 and will be removed in 1.0.0. An updated version of the class exists in the `langchain-ollama package and should be used instead. To use it run `pip install -U `langchain-ollama` and import as `from `langchain_ollama import OllamaLLM``.
  llm = Ollama(model="llama3.1")


["That's an easy one!", '', 'The answer to 2 + 2 is... 4!']

In [6]:
llm.invoke("Hello, world!")

"Hello! It's nice to meet you. Is there something I can help you with or would you like to chat?"

In [7]:
# openai = OpenAI()

# response = openai.chat.completions.create(model="gpt-5-nano", messages=messages)
# response.choices[0].message.content

## OK onwards with our first project

In [8]:
# Let's try out this utility

ed = fetch_website_contents("https://edwarddonner.com")
print(ed)

Home - Edward Donner

Home
Connect Four
Outsmart
An arena that pits LLMs against each other in a battle of diplomacy and deviousness
About
Posts
Well, hi there.
I’m Ed. I like writing code and experimenting with LLMs, and hopefully you’re here because you do too. I also enjoy DJing (but I’m badly out of practice), amateur electronic music production (
very
amateur) and losing myself in
Hacker News
, nodding my head sagely to things I only half understand.
I’m the co-founder and CTO of
Nebula.io
. We’re applying AI to a field where it can make a massive, positive impact: helping people discover their potential and pursue their reason for being. Recruiters use our product today to source, understand, engage and manage talent. I’m previously the founder and CEO of AI startup untapt,
acquired in 2021
.
We work with groundbreaking, proprietary LLMs verticalized for talent, we’ve
patented
our matching model, and our award-winning platform has happy customers and tons of press coverage.
Conne

## Types of prompts

You may know this already - but if not, you will get very familiar with it!

Models like GPT have been trained to receive instructions in a particular way.

They expect to receive:

**A system prompt** that tells them what task they are performing and what tone they should use

**A user prompt** -- the conversation starter that they should reply to

In [42]:
# Define our system prompt - you can experiment with this later, changing the last sentence to 'Respond in markdown in Spanish."

system_prompt = """
You are a helpful assistant that analyzes the contents of a website,
and provides the price of the product on the page. There might be other prices listed, but look for the price that are not strikedthrough
"""

In [50]:
# Define our user prompt

user_prompt_prefix = """
Here is the HTML content of a product's page. Please find the price of the product and respond with the price. The following is the HTML content of the page:

"""

## Messages

The API from OpenAI expects to receive messages in a particular structure.
Many of the other APIs share this structure:

```python
[
    {"role": "system", "content": "system message goes here"},
    {"role": "user", "content": "user message goes here"}
]
```
To give you a preview, the next 2 cells make a rather simple call - we won't stretch the mighty GPT (yet!)

In [51]:
messages = [
    {"role": "system", "content": "You are a helpful assistant"},
    {"role": "user", "content": "What is 2 + 2?"}
]

# response = llm.chat.completions.create(model="gpt-4.1-nano", messages=messages)
# response.choices[0].message.content

llm.invoke(input = messages)

'The answer to 2 + 2 is... 4!'

## And now let's build useful messages for GPT-4.1-mini, using a function

In [52]:
# See how this function creates exactly the format above

def messages_for(website):
    return [
        {"role": "system", "content": system_prompt},
        {"role": "user", "content": user_prompt_prefix + website}
    ]

## Time to bring it together - the API for OpenAI is very simple!

In [53]:

# URL of the page you want to fetch
url = "https://www.walmart.com/ip/Tylenol-Extra-Strength-500-mg-Acetaminophen-Rapid-Release-Gels-100-Ct/156205102?classType=REGULAR&athbdg=L1200&adsRedirect=true"
url = "https://www.cvs.com/shop/tylenol-extra-strength-caplets-50-ct-prodid-658932"
import requests
from bs4 import BeautifulSoup
import time
import random

def scrape_with_headers(url):
    """Method 1: Add realistic headers"""
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.5',
        'Accept-Encoding': 'gzip, deflate, br',
        'DNT': '1',
        'Connection': 'keep-alive',
        'Upgrade-Insecure-Requests': '1',
        'Referer': 'https://www.google.com/'
    }
    
    try:
        response = requests.get(url, headers=headers, timeout=10)
        response.raise_for_status()
        soup = BeautifulSoup(response.content, 'html.parser')
        return soup.prettify()
    except requests.exceptions.RequestException as e:
        print(f"Error with headers method: {e}")
        return None


def scrape_with_session(url):
    """Method 2: Use a session to maintain cookies"""
    session = requests.Session()
    
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
    }
    
    session.headers.update(headers)
    
    try:
        response = session.get(url, timeout=10)
        response.raise_for_status()
        return BeautifulSoup(response.content, 'html.parser')
    except requests.exceptions.RequestException as e:
        print(f"Error with session method: {e}")
        return None


def scrape_with_delay(urls):
    """Method 3: Add delays between requests"""
    results = []
    
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
    }
    
    for url in urls:
        try:
            # Random delay between 2-5 seconds
            time.sleep(random.uniform(2, 5))
            
            response = requests.get(url, headers=headers, timeout=10)
            response.raise_for_status()
            soup = BeautifulSoup(response.content, 'html.parser')
            results.append(soup)
            print(f"Successfully scraped: {url}")
            
        except requests.exceptions.RequestException as e:
            print(f"Error scraping {url}: {e}")
    
    return results


def scrape_with_selenium(url):
    """Method 4: Use Selenium for JavaScript-heavy sites
    Requires: pip install selenium
    Also need to download ChromeDriver"""
    
    try:
        from selenium import webdriver
        from selenium.webdriver.chrome.options import Options
        from selenium.webdriver.chrome.service import Service
        from selenium.webdriver.common.by import By
        from selenium.webdriver.support.ui import WebDriverWait
        from selenium.webdriver.support import expected_conditions as EC
        
        # Setup Chrome options
        chrome_options = Options()
        chrome_options.add_argument('--headless')  # Run in background
        chrome_options.add_argument('--no-sandbox')
        chrome_options.add_argument('--disable-dev-shm-usage')
        chrome_options.add_argument('user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36')
        
        # Initialize driver
        driver = webdriver.Chrome(options=chrome_options)
        
        # Load the page
        driver.get(url)
        
        # Wait for page to load (adjust selector as needed)
        WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.TAG_NAME, "body"))
        )
        
        # Get page source
        html = driver.page_source
        soup = BeautifulSoup(html, 'html.parser')
        
        driver.quit()
        return soup
        
    except ImportError:
        print("Selenium not installed. Run: pip install selenium")
        return None
    except Exception as e:
        print(f"Error with Selenium: {e}")
        return None


def scrape_with_retry(url, max_retries=3):
    """Method 5: Retry with exponential backoff"""
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
    }
    
    for attempt in range(max_retries):
        try:
            response = requests.get(url, headers=headers, timeout=10)
            response.raise_for_status()
            return BeautifulSoup(response.content, 'html.parser')
        except requests.exceptions.RequestException as e:
            if attempt < max_retries - 1:
                wait_time = 2 ** attempt  # Exponential backoff: 1, 2, 4 seconds
                print(f"Attempt {attempt + 1} failed. Retrying in {wait_time}s...")
                time.sleep(wait_time)
            else:
                print(f"All {max_retries} attempts failed: {e}")
                return None


# Example usage
if __name__ == "__main__":
    
    print("Trying Method 1: With Headers")
    soup = scrape_with_headers(url)
    if soup:
        print(soup.prettify())
        print(f"Success! Title: {soup.title.string if soup.title else 'No title'}")
    
    # print("\nTrying Method 2: With Session")
    # soup = scrape_with_session(url)
    # if soup:
    #     print(f"Success! Title: {soup.title.string if soup.title else 'No title'}")
    
    # print("\nTrying Method 5: With Retry")
    # soup = scrape_with_retry(url)
    # if soup:
    #     print(f"Success! Title: {soup.title.string if soup.title else 'No title'}")

Trying Method 1: With Headers
Error with headers method: 403 Client Error: Forbidden for url: https://www.cvs.com/shop/tylenol-extra-strength-caplets-50-ct-prodid-658932


In [54]:
from bs4 import BeautifulSoup
import requests
# url = "https://www.amazon.com/dp/B0BMX8N6XB?ref=ppx_yo2ov_dt_b_fed_asin_title&th=1"
url = "https://www.walmart.com/ip/Tylenol-Extra-Strength-500-mg-Acetaminophen-Rapid-Release-Gels-100-Ct/156205102"
# Standard headers to fetch a website
headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36"
}
response = requests.get(url, headers=headers)
soup = BeautifulSoup(response.content, "html.parser")

title = soup.title.string if soup.title else "No title found"
if soup.body:
    for irrelevant in soup.body(["script", "style", "img", "input"]):
        irrelevant.decompose()
    text = soup.body.get_text(separator="\n", strip=True)
else:
    text = ""

print((title + "\n\n" + text)[:2_000])

Robot or human?

Robot or human?
Activate and hold the button to confirm that you’re human. Thank You!
Terms of Use
Privacy Policy
Do Not Sell My Personal Information
Request My Personal Information
©
Walmart Stores, Inc.


In [ ]:
def get_price(url):
    website = scrape_with_headers(url)
    response = llm.invoke(
        input = messages_for(website)
    )
    return response

In [56]:
get_price("https://www.walmart.com/ip/Tylenol-Extra-Strength-500-mg-Acetaminophen-Rapid-Release-Gels-100-Ct/156205102")

'This is a HTML document that contains several JavaScript scripts. Here\'s an analysis of the content:\n\n**Scripts**\n\nThere are two main scripts embedded in this HTML document:\n\n1. The first script is a JSON object that contains various settings and configuration data for the application. This script has no `src` attribute, which means it\'s being generated by the server-side code.\n2. The second script is an empty script with the ID "release-metadata". It\'s likely used to store metadata about the application release.\n\n**JSON Object**\n\nThe first script contains a JSON object that holds various settings and configuration data for the application. Some of the key values in this object include:\n\n* `appVersion`: The version number of the application (usweb-1.231.3-62d525b7b0c34941587ba19e1e5d087f1a297d0e-O60116r)\n* `dataCenter`: The data center where the application is hosted (wus2)\n* Various other settings, such as `oidcParams`, `threatMetrix`, and `membershipConfig`\n\n**Se

In [20]:
# A function to display this nicely in the output, using markdown

def display_summary(url):
    summary = summarize(url)
    display(Markdown(summary))

In [21]:
display_summary("https://edwarddonner.com")

After analyzing the HTML content, I found that there is no price listed for any product on this page. The website appears to be a blog or personal website of Edward Donner, with various articles and posts about AI and related topics. There are no e-commerce elements or prices mentioned anywhere on the page.

If you would like, I can help you analyze another website or provide assistance with something else!

# Let's try more websites

Note that this will only work on websites that can be scraped using this simplistic approach.

Websites that are rendered with Javascript, like React apps, won't show up. See the community-contributions folder for a Selenium implementation that gets around this. You'll need to read up on installing Selenium (ask ChatGPT!)

Also Websites protected with CloudFront (and similar) may give 403 errors - many thanks Andy J for pointing this out.

But many websites will work just fine!

In [22]:
display_summary("https://en.wikipedia.org/wiki/Boeing_737_MAX_groundings")

I think there's been a mistake!

The provided HTML content appears to be the source code of an article on Wikipedia about Boeing 737 MAX groundings. However, I couldn't find any price information for a product.

There are no prices mentioned in this text, and it seems to be a general information article rather than a product page. If you could provide me with the actual HTML content of a product's page or give me more context about what specific product you're looking for, I'd be happy to help!

In [17]:
display_summary("https://cnn.com")

**CNN: Because You Clearly Can't Get Enough News**

The website appears to be a news aggregation platform, with an overwhelming amount of categories and subcategories. The main sections include US, World, Politics, Business, Health, Entertainment, and more. It seems they're trying to cover everything from the Ukraine-Russia War to celebrity news. They've also got a section called "Underscored" which is probably their attempt at being trendy and relevant.

**Breaking News**

There doesn't seem to be any actual breaking news on the website, just a call for feedback on ads. They're asking users how annoying they found the ads and whether they encountered any technical issues. Because who needs actual journalism when you can ask people about their ad experiences?

**Other Features**

The website also has a bunch of sections like "Watch" and "Listen", which probably link to video and audio content. There's also a section called "CNN Underscored" that seems to be focused on reviews, deals, and gifts. Because who doesn't love clicking through endless product recommendations?

In [18]:
display_summary("https://anthropic.com")

**This website is a snooze-fest.**

Anthropic is a "public benefit corporation" (read: corporate buzzword soup) that claims to care about AI safety while pushing out models like Claude, which sounds suspiciously like a fancy calculator. They're proud of their latest creation, Claude 4.5, which they tout as the "best model in the world." Because, you know, one size fits all and what could possibly go wrong with that? They're also big on transparency (right), initiatives (more buzzwords), and commitments (vague promises). In short, this website is a corporate brochure with a dash of AI jargon. Yay.

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/business.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#181;">Business applications</h2>
            <span style="color:#181;">In this exercise, you experienced calling the Cloud API of a Frontier Model (a leading model at the frontier of AI) for the first time. We will be using APIs like OpenAI at many stages in the course, in addition to building our own LLMs.

More specifically, we've applied this to Summarization - a classic Gen AI use case to make a summary. This can be applied to any business vertical - summarizing the news, summarizing financial performance, summarizing a resume in a cover letter - the applications are limitless. Consider how you could apply Summarization in your business, and try prototyping a solution.</span>
        </td>
    </tr>
</table>

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/important.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#900;">Before you continue - now try yourself</h2>
            <span style="color:#900;">Use the cell below to make your own simple commercial example. Stick with the summarization use case for now. Here's an idea: write something that will take the contents of an email, and will suggest an appropriate short subject line for the email. That's the kind of feature that might be built into a commercial email tool.</span>
        </td>
    </tr>
</table>

## An extra exercise for those who enjoy web scraping

You may notice that if you try `display_summary("https://openai.com")` - it doesn't work! That's because OpenAI has a fancy website that uses Javascript. There are many ways around this that some of you might be familiar with. For example, Selenium is a hugely popular framework that runs a browser behind the scenes, renders the page, and allows you to query it. If you have experience with Selenium, Playwright or similar, then feel free to improve the Website class to use them. In the community-contributions folder, you'll find an example Selenium solution from a student (thank you!)

# Sharing your code

I'd love it if you share your code afterwards so I can share it with others! You'll notice that some students have already made changes (including a Selenium implementation) which you will find in the community-contributions folder. If you'd like add your changes to that folder, submit a Pull Request with your new versions in that folder and I'll merge your changes.

If you're not an expert with git (and I am not!) then GPT has given some nice instructions on how to submit a Pull Request. It's a bit of an involved process, but once you've done it once it's pretty clear. As a pro-tip: it's best if you clear the outputs of your Jupyter notebooks (Edit >> Clean outputs of all cells, and then Save) for clean notebooks.

Here are good instructions courtesy of an AI friend:  
https://chatgpt.com/share/677a9cb5-c64c-8012-99e0-e06e88afd293